### ISSUE TITLE
Creating 1155 Collection with Privy Provider

### ISSUE THREAD

**Issue by mtple** [ourzora/ocs-hackathon-24#6]
Hello, I am trying to deploy a new 1155 collection from a Privy embedded wallet and am getting this error:

Error creating collection: [TypeError: publicClient.readContract is not a function (it is undefined)]

Based on that error and looking at the SDK code, it appears that the SDK for creating collections only works with Viem providers, which is the only one with the readContract function as far as I can tell. Is this the case, and if so, is it still possible to create collections from a Privy Provider? FYI, Privy Provider follows the EIP 1193 standard.

**Comment by oveddan** [comment]
hey @mattleesounds could you provide a more full code example?

**Comment by mattleesounds** [comment]
Sure, here ya go

`import React, { useState } from "react"
import { Pressable, Text } from "react-native"
import { useEmbeddedWallet, isConnected } from "@privy-io/expo"
import { create1155CreatorClient } from "@zoralabs/protocol-sdk"

const CreateCollection: React.FC = () => {
  const wallet = useEmbeddedWallet()
  const [isCreated, setIsCreated] = useState(false)

  const createCollection = async (provider: EIP1193Provider) => {
    try {
      const accounts = await provider.request({
        method: "eth_requestAccounts",
      })
      const creatorClient = create1155CreatorClient({ publicClient: provider })
      const { request } = await creatorClient.createNew1155Token({
        contract: {
          name: "test-collection-1",
          uri: "https://gold-extraordinary-pigeon-404.mypinata.cloud/ipfs/QmbVrAKuxiHDVEdzhkQcC5euxYm3sLuf4M25FmavQgDpzm",
        },
        tokenMetadataURI:
          "https://gold-extraordinary-pigeon-404.mypinata.cloud/ipfs/QmXxJHtGti9r2jLpKkRDmQqnMkmb6DHUnUZd9RuvxPnHe5",
        account: accounts[0],
        mintToCreatorCount: 1,
      })
      const { request: simulateRequest } = provider.simulateContract(request)
      const hash = await provider.send({
        method: "eth_sendTransaction",
        params: [simulateRequest],
      })
      const receipt = await provider.request({
        method: "eth_getTransactionReceipt",
        params: [hash],
      })
      setIsCreated(true)
      console.log("Collection created:", receipt)
    } catch (error) {
      setIsCreated(false)
      console.error("Error creating collection:", error)
    }
  }

  return (
    <>
      <Pressable
        className="pt-4 "
        onPress={() => isConnected(wallet) && createCollection(wallet.provider)}
      >
        <Text className="text-white border-2  border-white text-center text-xl font-bold">
          CREATE COLLECTION
        </Text>
      </Pressable>
      {isCreated && (
        <Text className="text-green-500">Collection successfully created!</Text>
      )}
    </>
  )
}

export default CreateCollection
`

**Comment by oveddan** [comment]
hey so i tried to repro this locally. i was getting a typescript error at 
```ts
onPress={() => isConnected(wallet) && createCollection(wallet.provider)}
```
that the type didn't match `EIP1193Provider`

it looks like `wallet.provider` is a `PrivyEmbeddedWalletProvider` so I tried plugging that into the argument, and `readContract` is not defined:

<img width="595" alt="Screenshot 2024-06-07 at 12 11 39 PM" src="https://github.com/ourzora/ocs-hackathon-24/assets/891755/4c03e2f1-4bb2-44df-8020-12c453373533">


we could add a better example and support for integration with privy. for now I found in their docs examples of creating a publicClient for using with other libs:

https://docs.privy.io/guide/react/recipes/account-abstraction/pimlico#_4-create-a-smart-account-for-your-user
https://docs.privy.io/guide/react/recipes/account-abstraction/wagmi#_1-initialize-an-eip1193-provider-for-the-smart-account

Perhaps until we better support privy you could manually setup a PublicClient like those example?

**Comment by mattleesounds** [comment]
Will attempt! Thanks for the privy resources.

**Comment by Nith567** [comment]
@oveddan  Hey i am trying to mint nft's i saw the abi : "inputs": [
        {
            "internalType": "contract IMinter1155",
            "name": "minter",
            "type": "address"
        },
        {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256"
        },
        {
            "internalType": "uint256",
            "name": "quantity",
            "type": "uint256"
        },
        {
            "internalType": "address[]",
            "name": "rewardsRecipients",
            "type": "address[]"
        },
        {
            "internalType": "bytes",
            "name": "minterArguments",
            "type": "bytes"
        }
    ],  here what are the minterArguments contains  encodedbytes ? i saw in a sdk that it has :     const mintArguments = {
      quantityToMint: 1,
      mintComment: "Summer Chain!",
      mintReferral: "0x8879318091671ba1274e751f8cdef76bb37eb3ed"
    };
 but in args we already have quantitytomint so can you confirm it, while minting facing few errros thats why

**Comment by Nith567** [comment]
@oveddan check this out: https://github.com/Nith567/mint-/blob/main/src/components/Transact.tsx
when i try to mint it, says: (execution reverted with data Try again)

**Comment by mattleesounds** [comment]
@oveddan FYI, this turned out to be the relevant docs for integration with privy-expo SDK [](https://docs.privy.io/guide/expo/embedded/3p-libraries)

**Comment by iainnash** [comment]
Hi!

We have different minter modules that take different arguments. Our SDK takes those arguments and encodes them for the fixed price minter module (our default).


It looks like the reason the last few transactions failed is that you attempted to call an NFT contract that has no NFTs/Mints registered: 
<img width="1037" alt="Screenshot 2024-06-10 at 11 50 53 AM" src="https://github.com/ourzora/ocs-hackathon-24/assets/835896/105f7f53-75db-4aca-a449-d69a21b940bf">

**Comment by iainnash** [comment]
Can you try with an NFT that has a token created/registered on it with a sales strategy?

**Comment by oveddan** [comment]
> @oveddan FYI, this turned out to be the relevant docs for integration with privy-expo SDK https://docs.privy.io/guide/expo/embedded/3p-libraries

great link @mattleesounds thank you! we can work on adding a section on privy integration to the docs, or look at ways to have more native integration

**Comment by Nith567** [comment]
@iainnash yeah it did solved after i put correct parameters, tried through abi using wagmi hooks it worked out instead of sdk.
